core/dnsserver/server_quic.go

```diff
@@ -0,0 +1,293 @@
+func (s *ServerQUIC) ServePacket(p net.PacketConn) error {
+	s.m.Lock()
+	s.listenAddr = s.quicListener.Addr()
+	s.m.Unlock()
+
+	return s.ServeQUIC()
+}
+
+// ServeQUIC listens for incoming QUIC packets.
+func (s *ServerQUIC) ServeQUIC() error {
+	for {
+		conn, err := s.quicListener.Accept(context.Background())
+		if err != nil {
+			closeQUICConn(conn, DoQCodeNoError)
+			return err
+		}
+
+		go s.serveQUICConnection(conn)
+	}
+}
+
+// serveQUICConnection handles a new QUIC connection. It waits for new streams
+// and passes them to serveQUICStream.
+func (s *ServerQUIC) serveQUICConnection(conn quic.Connection) {
+	for {
+		// In DoQ, one query consumes one stream.
+		// The client MUST select the next available client-initiated bidirectional
+		// stream for each subsequent query on a QUIC connection.
+		stream, err := conn.AcceptStream(context.Background())
+		if err != nil {
+			closeQUICConn(conn, DoQCodeNoError)
+			return
+		}
+
+		go func() {
+			s.serveQUICStream(stream, conn)
+
+			// The server MUST send the response(s) on the same stream and MUST
+			// indicate, after the last response, through the STREAM FIN
+			// mechanism that no further data will be sent on that stream.
+			_ = stream.Close()
+		}()
+
+	}
+}
+
+func (s *ServerQUIC) serveQUICStream(stream quic.Stream, conn quic.Connection) {
+	bufPtr := s.bytesPool.Get().(*[]byte)
+	defer s.bytesPool.Put(bufPtr)
+
+	buf := *bufPtr
+	_, err := readAll(stream, buf)
+
+	// io.EOF does not really mean that there's any error, it is just
+	// the STREAM FIN indicating that there will be no data to read
+	// anymore from this stream.
+	if err != nil && err != io.EOF {
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	req := &dns.Msg{}
+
+	// Drafts of the RFC9250 did not require the 2-byte prefixed message length.
+	// Thus, we are only supporting the official version (DoQ v1).
+	err = req.Unpack(buf[2:])
+	if err != nil {
+		clog.Debug("unpacking quic packet: %s", err)
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	if !validRequest(req) {
+		// If a peer encounters such an error condition, it is considered a
+		// fatal error. It SHOULD forcibly abort the connection using QUIC's
+		// CONNECTION_CLOSE mechanism and SHOULD use the DoQ error code
+		// DOQ_PROTOCOL_ERROR.
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	w := &DoQWriter{
+		localAddr:  conn.LocalAddr(),
+		remoteAddr: conn.RemoteAddr(),
+		stream:     stream,
+		Msg:        req,
+	}
+
+	dnsCtx := context.WithValue(context.Background(), Key{}, s.Server)
+	dnsCtx = context.WithValue(dnsCtx, LoopKey{}, 0)
+	s.ServeDNS(dnsCtx, w, req)
+}
+
+// ListenPacket implements caddy.UDPServer interface.
+func (s *ServerQUIC) ListenPacket() (net.PacketConn, error) {
+	p, err := reuseport.ListenPacket("udp", s.Addr[len(transport.QUIC+"://"):])
+	if err != nil {
+		return nil, err
+	}
+
+	s.quicListener, err = quic.Listen(p, s.tlsConfig, s.quicConfig)
```

Comment:
To be honest, I don't really understand the necessity of these mutex locks. I asked Miek about it and it looks like it is needed to avoid data races when shutting down the server. 

Previously, I put them according to the other servers (`server_grpc and `server_https`, mainly) and none of them protect the listener initialization.

Anyway, it does make sense to do it, so I protected it as you mentioned.